File: tooling/python/labkit/labkit/cache.py
"""
Redis cache / coordination layer for the labs.

Connection comes from REDIS_URL, pre-injected by the lab platform.
Learners just `from labkit import cache`.
"""

import json
import os

import redis


def _client() -> redis.Redis:
    url = os.environ.get("REDIS_URL", "redis://localhost:6379/0")
    return redis.from_url(url, decode_responses=True)


class Cache:
    """Thin, ready-to-use Redis handle with JSON helpers."""

    def __init__(self):
        self._client = None

    @property
    def client(self) -> redis.Redis:
        if self._client is None:
            self._client = _client()
        return self._client

    def get(self, key: str) -> str | None:
        return self.client.get(key)

    def set(self, key: str, value: str, ttl: int | None = None):
        self.client.set(key, value, ex=ttl)

    def get_json(self, key: str):
        raw = self.client.get(key)
        return json.loads(raw) if raw is not None else None

    def set_json(self, key: str, value, ttl: int | None = None):
        self.client.set(key, json.dumps(value), ex=ttl)

    def delete(self, *keys: str) -> int:
        if not keys:
            return 0
        return self.client.delete(*keys)

    def exists(self, key: str) -> bool:
        return bool(self.client.exists(key))

    def incr(self, key: str, amount: int = 1) -> int:
        return self.client.incr(key, amount)

    def flush(self):
        """Clear the cache (test isolation)."""
        self.client.flushdb()

    def ping(self) -> bool:
        try:
            return bool(self.client.ping())
        except Exception:
            return False


cache = Cache()

File: tooling/go/labkit/cache.go
package labkit

import (
	"context"
	"encoding/json"
	"os"
	"time"

	"github.com/redis/go-redis/v9"
)

type cacheHandle struct {
	client *redis.Client
	ctx    context.Context
}

func newCache() *cacheHandle {
	url := os.Getenv("REDIS_URL")
	if url == "" {
		url = "redis://localhost:6379/0"
	}
	opt, err := redis.ParseURL(url)
	if err != nil {
		panic(err)
	}
	return &cacheHandle{client: redis.NewClient(opt), ctx: context.Background()}
}

// GetJSON unmarshals the cached value for key into dest. Returns false on miss.
func (c *cacheHandle) GetJSON(key string, dest any) bool {
	val, err := c.client.Get(c.ctx, key).Result()
	if err == redis.Nil {
		return false
	}
	if err != nil {
		panic(err)
	}
	if err := json.Unmarshal([]byte(val), dest); err != nil {
		panic(err)
	}
	return true
}

// SetJSON stores value as JSON with an optional TTL (seconds; 0 = no expiry).
func (c *cacheHandle) SetJSON(key string, value any, ttlSeconds int) {
	b, err := json.Marshal(value)
	if err != nil {
		panic(err)
	}
	var exp time.Duration
	if ttlSeconds > 0 {
		exp = time.Duration(ttlSeconds) * time.Second
	}
	if err := c.client.Set(c.ctx, key, b, exp).Err(); err != nil {
		panic(err)
	}
}

func (c *cacheHandle) Delete(keys ...string) {
	if len(keys) > 0 {
		c.client.Del(c.ctx, keys...)
	}
}

func (c *cacheHandle) Exists(key string) bool {
	n, _ := c.client.Exists(c.ctx, key).Result()
	return n > 0
}

func (c *cacheHandle) Flush()    { c.client.FlushDB(c.ctx) }
func (c *cacheHandle) Ping() bool { return c.client.Ping(c.ctx).Err() == nil }

// Cache is the ready Redis handle.
var Cache = newCache()
